File: data-structures/dynamic-ring-buffer/implementations/dynamic-ring-buffer.ts
import { DynamicRingBuffer } from "../interfaces/dynamic-ring-buffer";

export class MyDynamicRingBuffer<T> implements DynamicRingBuffer<T> {
  private readonly initialCapacity: number;
  private currentCapacity: number
  private head: number
  private tail: number
  private container: Array<T>

  // bookeeping
  private currentLength: number

  public constructor(initialCapacity = 4) {
    if (initialCapacity <= 0) {
      throw new Error("initialCapacity must be greater than zero");
    }

    this.initialCapacity = initialCapacity;
    this.currentCapacity = initialCapacity
    this.container = new Array(initialCapacity)
    this.head = 0;
    this.tail = 0
    this.currentLength = 0
  }

  public length(): number {
    return this.currentLength
  }

  public capacity(): number {
    return this.currentCapacity
  }

  public isEmpty(): boolean {
    return this.length() <= 0
  }

  public isFull(): boolean {
    return this.length() === this.capacity()
  }

  public front(): T | undefined {
    if(this.isEmpty()) return undefined
    return this.container[this.head]
  }

  public rear(): T | undefined {
    if(this.isEmpty()) return undefined
    const idx = (this.tail - 1 + this.capacity()) % this.capacity()
    return this.container[idx]
  }

  public enqueue(_value: T): boolean {
    this.ensureCapacity()
    this.container[this.tail] = _value
    this.currentLength++
    this.tail = (this.tail + 1) % this.capacity()
    return true
  }

  public dequeue(): T | undefined {
    if(this.isEmpty()) return undefined;
    const val = this.container[this.head]
    this.head = (this.head + 1) % this.capacity()
    this.currentLength--
    return val
  }

  public clear(): void {
    this.currentCapacity = this.getInitialCapacity()
    this.container = new Array(this.getInitialCapacity())
    this.head = 0;
    this.tail = 0
    this.currentLength = 0
  }

  public toArray(): T[] {
    const result = [] as T[]

    for(let i = 0; i < this.length(); i++) {
      const idx = (this.head + i) % this.capacity()
      const v = this.container[idx]
      if(v && v !== undefined) {
        result.push(v)
      }
    }

    return result
  }

  protected getInitialCapacity(): number {
    return this.initialCapacity;
  }

  private ensureCapacity() {
    if(!this.isFull()) return

    const newCapacity = this.currentCapacity * 2
    const newContainer = new Array<T>(newCapacity)

    // copy backing array
    for(let i = 0; i < this.length(); i++) {
      const currIdx = (this.head + i) % this.capacity()
      const v = this.container[currIdx]
      newContainer[i] = v
    }

    this.head = 0
    this.tail = this.currentCapacity
    this.currentCapacity = newCapacity
    this.container = newContainer
  }
}

File: data-structures/dynamic-ring-buffer/implementations/dynamic-ring-buffer.go
package implementations

import "dsa/data-structures/dynamic-ring-buffer/interfaces"

type MyDynamicRingBuffer[T any] struct {
	initialCapacity int
	currentCapacity int
	currentLength   int
	head            int
	tail            int
	container       []T
}

var _ interfaces.DynamicRingBuffer[int] = (*MyDynamicRingBuffer[int])(nil)

func NewMyDynamicRingBuffer[T any](initialCapacity int) *MyDynamicRingBuffer[T] {
	if initialCapacity <= 0 {
		panic("initialCapacity must be greater than zero")
	}

	container := make([]T, initialCapacity)
	return &MyDynamicRingBuffer[T]{
		initialCapacity: initialCapacity,
		currentCapacity: initialCapacity,
		container:       container,

		// go's zero value will ultimately take care of these fields, but better to be explicit
		currentLength: 0,
		head:          0,
		tail:          0,
	}
}

func (b *MyDynamicRingBuffer[T]) Length() int {
	return b.currentLength
}

func (b *MyDynamicRingBuffer[T]) Capacity() int {
	return b.currentCapacity
}

func (b *MyDynamicRingBuffer[T]) IsEmpty() bool {
	return b.Length() <= 0
}

func (b *MyDynamicRingBuffer[T]) IsFull() bool {
	return b.Length() == b.Capacity()
}

func (b *MyDynamicRingBuffer[T]) Front() T {
	if b.IsEmpty() {
		panic("buffer is empty")
	}

	return b.container[b.head]
}

func (b *MyDynamicRingBuffer[T]) Rear() T {
	if b.IsEmpty() {
		panic("buffer is empty")
	}

	idx := (b.tail - 1 + b.Capacity()) % b.Capacity()
	return b.container[idx]
}

func (b *MyDynamicRingBuffer[T]) Enqueue(value T) bool {
	b.ensureCapacity()
	b.container[b.tail] = value
	b.currentLength++
	b.tail = (b.tail + 1) % b.Capacity()
	return true
}

func (b *MyDynamicRingBuffer[T]) Dequeue() T {
	if b.IsEmpty() {
		panic("buffer is empty")
	}

	val := b.container[b.head]
	b.head = (b.head + 1) % b.Capacity()
	b.currentLength--
	return val
}

func (b *MyDynamicRingBuffer[T]) Clear() {
	b.currentLength = 0
	b.head = 0
	b.tail = 0
	b.container = nil
}

func (b *MyDynamicRingBuffer[T]) ToSlice() []T {
	r := make([]T, 0, b.Length())

	for i := 0; i < b.Length(); i++ {
		idx := (b.head + i) % b.Capacity()
		v := b.container[idx]
		r = append(r, v)
	}

	return r
}

func (b *MyDynamicRingBuffer[T]) ensureCapacity() {
	if !b.IsFull() {
		return
	}

	newCap := b.currentCapacity * 2
	newCon := make([]T, newCap)

	for i := 0; i < b.Length(); i++ {
		idx := (b.head + i) % b.Capacity()
		v := b.container[idx]
		newCon[i] = v
	}

	b.head = 0
	b.tail = b.currentCapacity
	b.currentCapacity = newCap
	b.container = newCon

}

func (b *MyDynamicRingBuffer[T]) InitialCapacity() int {
	return b.initialCapacity
}
